File: impl/golang/internal/topic/service.go
package topic

import (
	"encoding/json"
	"fmt"
)

// HCSConfig represents the transport configuration for an HCS-backed topic service.
type HCSConfig struct {
	Network string `json:"network"`
	TopicId string `json:"topicId"`
}

// ERCLogConfig represents the transport configuration for an ERC event log topic service.
type ERCLogConfig struct {
	ChainId         uint64 `json:"chainId"`
	ContractAddress string `json:"contractAddress"`
	EventSignature  string `json:"eventSignature"`
}

// FR-T16: Anchoring for non-ledger transports
// AnchoringConfig defines how a non-ledger transport anchors its messages to a ledger.
type AnchoringConfig struct {
	Method        string `json:"method"`
	AnchorTopicId string `json:"anchorTopicId"`
	AnchorNetwork string `json:"anchorNetwork"`
	Interval      string `json:"interval"`
}

// KafkaConfig represents the transport configuration for a Kafka-backed topic service.
type KafkaConfig struct {
	BootstrapServers []string        `json:"bootstrapServers"`
	TopicName        string          `json:"topicName"`
	SASLMechanism    string          `json:"saslMechanism,omitempty"`
	Anchoring        AnchoringConfig `json:"anchoring"`
}

// FR-T14: EIP-8004 service representation (type: neuron-topic)
// NeuronTopicServiceDef is the Spec 004 representation of a neuron-topic service entry
// as found in an agentURI JSON document. It captures the typed transport configuration
// and channel binding for a single topic service.
type NeuronTopicServiceDef struct {
	Type      string      `json:"type"`
	Name      string      `json:"name"`
	Endpoint  string      `json:"endpoint,omitempty"`
	Version   string      `json:"version"`
	Channel   string      `json:"channel"`
	Transport string      `json:"transport"`
	Anchor    string      `json:"anchor"`
	Config    interface{} `json:"config"`
}

// NeuronP2PExchangeServiceDef is defined in p2p_exchange.go to avoid circular references.
// This forward declaration comment is for documentation only.

// agentURIDocument is a minimal representation of an agentURI JSON document.
// It is used only for parsing purposes; the full schema is defined in the registry package.
type agentURIDocument struct {
	Services []json.RawMessage `json:"services"`
}

// serviceTypeField is used to peek at the "type" field of a service entry.
type serviceTypeField struct {
	Type string `json:"type"`
}

// FR-T15: Transport config schemas (HCS, ERC-log, Kafka)
// ValidateTransportConfig validates a transport configuration map against the
// schema for the given transport type. Returns ErrInvalidConfig if required
// fields are missing.
func ValidateTransportConfig(transport string, config map[string]interface{}) error {
	switch transport {
	case "hcs":
		return validateHCSConfig(config)
	case "erc-log":
		return validateERCLogConfig(config)
	case "kafka":
		return validateKafkaConfig(config)
	default:
		return NewTopicError(ErrUnsupportedTransport,
			fmt.Sprintf("unknown transport for config validation: %s", transport))
	}
}

func validateHCSConfig(config map[string]interface{}) error {
	if _, ok := config["network"]; !ok {
		return NewTopicError(ErrInvalidConfig, "HCS config missing required field: network")
	}
	if _, ok := config["topicId"]; !ok {
		return NewTopicError(ErrInvalidConfig, "HCS config missing required field: topicId")
	}
	return nil
}

func validateERCLogConfig(config map[string]interface{}) error {
	if _, ok := config["chainId"]; !ok {
		return NewTopicError(ErrInvalidConfig, "ERC-log config missing required field: chainId")
	}
	if _, ok := config["contractAddress"]; !ok {
		return NewTopicError(ErrInvalidConfig, "ERC-log config missing required field: contractAddress")
	}
	if _, ok := config["eventSignature"]; !ok {
		return NewTopicError(ErrInvalidConfig, "ERC-log config missing required field: eventSignature")
	}
	return nil
}

func validateKafkaConfig(config map[string]interface{}) error {
	if _, ok := config["bootstrapServers"]; !ok {
		return NewTopicError(ErrInvalidConfig, "Kafka config missing required field: bootstrapServers")
	}
	if _, ok := config["topicName"]; !ok {
		return NewTopicError(ErrInvalidConfig, "Kafka config missing required field: topicName")
	}
	anchoring, ok := config["anchoring"]
	if !ok {
		return NewTopicError(ErrInvalidConfig, "Kafka config missing required field: anchoring")
	}
	// Validate the anchoring sub-object if it is a map.
	if anchorMap, ok := anchoring.(map[string]interface{}); ok {
		for _, field := range []string{"method", "anchorTopicId", "anchorNetwork", "interval"} {
			if _, ok := anchorMap[field]; !ok {
				return NewTopicError(ErrInvalidConfig,
					fmt.Sprintf("Kafka anchoring config missing required field: %s", field))
			}
		}
	}
	return nil
}

// SerializeTopicService serializes a NeuronTopicServiceDef to JSON.
func SerializeTopicService(svc NeuronTopicServiceDef) ([]byte, error) {
	return json.Marshal(svc)
}

// FR-T09: Topic discovery via Peer Registry
// ParseAgentURIServices parses an agentURI JSON document and extracts
// neuron-topic and neuron-p2p-exchange service entries.
// Returns separate slices for topic services and P2P exchange services.
func ParseAgentURIServices(jsonBytes []byte) ([]NeuronTopicServiceDef, []NeuronP2PExchangeServiceDef, error) {
	var doc agentURIDocument
	if err := json.Unmarshal(jsonBytes, &doc); err != nil {
		return nil, nil, WrapTopicError(ErrInvalidConfig, "failed to parse agentURI JSON", err)
	}

	var topicServices []NeuronTopicServiceDef
	var p2pServices []NeuronP2PExchangeServiceDef

	for _, raw := range doc.Services {
		var typeField serviceTypeField
		if err := json.Unmarshal(raw, &typeField); err != nil {
			continue // Skip entries that cannot be parsed.
		}

		switch typeField.Type {
		case "neuron-topic":
			var svc NeuronTopicServiceDef
			if err := json.Unmarshal(raw, &svc); err != nil {
				continue
			}
			topicServices = append(topicServices, svc)
		case "neuron-p2p-exchange":
			var svc NeuronP2PExchangeServiceDef
			if err := json.Unmarshal(raw, &svc); err != nil {
				continue
			}
			p2pServices = append(p2pServices, svc)
		}
	}

	return topicServices, p2pServices, nil
}

// ExtractTopicRef extracts a TopicRef from a NeuronTopicServiceDef.
// It uses the transport field to determine the BackendKind, then builds
// a locator from the typed config.
func ExtractTopicRef(svc NeuronTopicServiceDef) (TopicRef, error) {
	transport, err := ParseBackendKind(svc.Transport)
	if err != nil {
		return TopicRef{}, err
	}

	// Re-marshal the config to a map for uniform field extraction.
	configBytes, err := json.Marshal(svc.Config)
	if err != nil {
		return TopicRef{}, WrapTopicError(ErrInvalidConfig, "failed to marshal service config", err)
	}

	var configMap map[string]interface{}
	if err := json.Unmarshal(configBytes, &configMap); err != nil {
		return TopicRef{}, WrapTopicError(ErrInvalidConfig, "service config is not a JSON object", err)
	}

	var locator string
	switch transport {
	case BackendHCS:
		topicId, ok := configMap["topicId"]
		if !ok {
			return TopicRef{}, NewTopicError(ErrInvalidConfig, "HCS config missing topicId")
		}
		locator = fmt.Sprintf("%v", topicId)
	case BackendERCLog:
		addr, ok := configMap["contractAddress"]
		if !ok {
			return TopicRef{}, NewTopicError(ErrInvalidConfig, "ERC-log config missing contractAddress")
		}
		locator = fmt.Sprintf("%v", addr)
	case BackendKafka:
		name, ok := configMap["topicName"]
		if !ok {
			return TopicRef{}, NewTopicError(ErrInvalidConfig, "Kafka config missing topicName")
		}
		locator = fmt.Sprintf("%v", name)
	default:
		return TopicRef{}, NewTopicError(ErrUnsupportedTransport,
			fmt.Sprintf("cannot extract locator for transport: %s", transport))
	}

	return NewTopicRef(transport, locator)
}

// DiscoverPeerChannels parses an agentURI JSON document and returns a map of
// standard channel roles to their TopicRef. It extracts all neuron-topic services
// and maps the ones with standard channel names (stdIn, stdOut, stdErr).
func DiscoverPeerChannels(agentURIJson []byte) (map[ChannelRole]TopicRef, error) {
	topicServices, _, err := ParseAgentURIServices(agentURIJson)
	if err != nil {
		return nil, err
	}

	result := make(map[ChannelRole]TopicRef)
	for _, svc := range topicServices {
		role, err := ChannelRoleFromString(svc.Channel)
		if err != nil {
			// Skip non-standard channels or invalid names.
			continue
		}

		ref, err := ExtractTopicRef(svc)
		if err != nil {
			continue
		}

		// Only include standard channel roles in the discovery result.
		if !IsCustomChannel(role) {
			result[role] = ref
		}
	}

	return result, nil
}

// FindStdIn finds the first neuron-topic service with channel "stdIn".
func FindStdIn(services []NeuronTopicServiceDef) (*NeuronTopicServiceDef, error) {
	return findServiceByChannel(services, string(ChannelStdIn))
}

// FindStdOut finds the first neuron-topic service with channel "stdOut".
func FindStdOut(services []NeuronTopicServiceDef) (*NeuronTopicServiceDef, error) {
	return findServiceByChannel(services, string(ChannelStdOut))
}

// FindStdErr finds the first neuron-topic service with channel "stdErr".
func FindStdErr(services []NeuronTopicServiceDef) (*NeuronTopicServiceDef, error) {
	return findServiceByChannel(services, string(ChannelStdErr))
}

// findServiceByChannel returns the first service with the matching channel name.
// Returns nil and ErrTopicNotFound if no matching service is found.
func findServiceByChannel(services []NeuronTopicServiceDef, channel string) (*NeuronTopicServiceDef, error) {
	for i := range services {
		if services[i].Channel == channel {
			return &services[i], nil
		}
	}
	return nil, NewTopicError(ErrTopicNotFound,
		fmt.Sprintf("no neuron-topic service found with channel: %s", channel))
}

File: impl/typescript/src/topic/service.ts
/**
 * Service schema types -- NeuronTopicService and NeuronP2PExchangeService.
 *
 * Spec reference: 004 spec.md
 *   - FR-T14: NeuronTopicService EIP-8004 service object.
 *   - FR-T15: Transport-specific configuration schemas.
 *   - FR-T17: NeuronP2PExchangeService for multiaddress discovery.
 *   - FR-T18: Cross-reference validation (topicRef resolves to existing service).
 *
 * Spec reference: 004 contracts/service-schema.md
 *   - Full JSON schemas for NeuronTopicService, NeuronP2PExchangeService.
 *   - Transport config schemas: HCSConfig, ERCLogConfig, KafkaConfig.
 *   - Parsing, validation, and serialization functions.
 *
 * SC-T09: Round-trip guarantee (serialize then parse produces identical result).
 * SC-T10: Cross-reference validation catches broken topicRef references.
 */

import type { BackendKind, ChannelRole } from './types.js';
import { isValidBackendKind, isValidChannelRole } from './types.js';
import { TopicRef } from './topic-ref.js';
import { invalidTopicRef } from './errors.js';

// ---------------------------------------------------------------------------
// Transport Config types
// ---------------------------------------------------------------------------

/**
 * HCS transport configuration.
 * FR-T15: Hedera Consensus Service backend.
 */
export interface HCSConfig {
  /** Hedera network identifier (e.g., 'hedera-mainnet', 'hedera-testnet'). FR-T15 */
  readonly network: string;

  /** HCS topic ID in shard.realm.num format (e.g., '0.0.4515382'). FR-T15 */
  readonly topicId: string;
}

/**
 * ERC event log transport configuration.
 * FR-T15: Read-only ERC event logs on Ethereum/EVM chains.
 */
export interface ERCLogConfig {
  /** EVM chain ID (e.g., 1 for Ethereum mainnet). FR-T15 */
  readonly chainId: number;

  /** EVM contract address (EIP-55 checksummed). FR-T15 */
  readonly contractAddress: string;

  /** Solidity event signature. FR-T15 */
  readonly eventSignature: string;
}

/**
 * Kafka anchoring configuration.
 * FR-T16: Non-ledger-native transports MUST have valid anchoring config.
 */
export interface AnchoringConfig {
  /** Anchoring method (e.g., 'hcs-hash-chain'). FR-T16 */
  readonly method: string;

  /** Topic ID on the anchor ledger. FR-T16 */
  readonly anchorTopicId: string;

  /** Network identifier of the anchor ledger. FR-T16 */
  readonly anchorNetwork: string;

  /** Anchoring frequency (e.g., 'every-batch', 'every-100-messages'). FR-T16 */
  readonly interval: string;
}

/**
 * Kafka transport configuration.
 * FR-T15, FR-T16: Kafka with ledger anchoring.
 */
export interface KafkaConfig {
  /** Kafka broker addresses. FR-T15 */
  readonly bootstrapServers: readonly string[];

  /** Kafka topic name. FR-T15 */
  readonly topicName: string;

  /** SASL authentication mechanism (optional). FR-T15 */
  readonly saslMechanism?: string | undefined;

  /** Ledger anchoring configuration. FR-T16 */
  readonly anchoring: AnchoringConfig;
}

/** Union of all transport config types. FR-T15 */
export type TransportConfig = HCSConfig | ERCLogConfig | KafkaConfig | Record<string, unknown>;

// ---------------------------------------------------------------------------
// NeuronTopicService
// ---------------------------------------------------------------------------

/**
 * EIP-8004 service object with type "neuron-topic".
 *
 * FR-T14: Represents a single topic channel in the agentURI services array.
 * All MUST fields are required by the JSON schema.
 */
export interface NeuronTopicService {
  /** Discriminator. Always 'neuron-topic'. FR-T14 */
  readonly type: 'neuron-topic';

  /** Channel role: stdIn, stdOut, stdErr, or custom:<name>. FR-T14, FR-T07, FR-T08 */
  readonly name: string;

  /** Compact Topic URI for backward-compatible EIP-8004 consumers. FR-T14 (SHOULD) */
  readonly endpoint?: string | undefined;

  /** Topic protocol version (semver). FR-T14 */
  readonly version: string;

  /** Neuron channel role (same as name for standard channels). FR-T14 */
  readonly channel: string;

  /** Backend kind. FR-T14, FR-T15 */
  readonly transport: string;

  /** Ledger that anchors the topic. FR-T14 */
  readonly anchor: string;

  /** Transport-specific configuration. FR-T14, FR-T15 */
  readonly config: TransportConfig;
}

// ---------------------------------------------------------------------------
// NeuronP2PExchangeService
// ---------------------------------------------------------------------------

/**
 * EIP-8004 service object with type "neuron-p2p-exchange".
 *
 * FR-T17: Defines the method for multiaddress discovery.
 * All MUST fields are required.
 */
export interface NeuronP2PExchangeService {
  /** Discriminator. Always 'neuron-p2p-exchange'. FR-T17 */
  readonly type: 'neuron-p2p-exchange';

  /** Service name (e.g., 'p2p'). FR-T17 */
  readonly name: string;

  /** Exchange protocol version (semver). FR-T17 */
  readonly version: string;

  /** Libp2p PeerID from Key Library 002. FR-T17 */
  readonly peerID: string;

  /** Protocol ID for multiaddress exchange. FR-T17 */
  readonly protocol: string;

  /**
   * Cross-reference to a neuron-topic service name in the same agentURI.
   * FR-T17, FR-T18: Must resolve to an existing neuron-topic service.
   */
  readonly topicRef: string;
}

// ---------------------------------------------------------------------------
// Parsing functions
// ---------------------------------------------------------------------------

/**
 * Parse a NeuronTopicService from a JSON object.
 *
 * FR-T14: Validates all required fields.
 * FR-T15: Validates transport-specific config structure.
 *
 * @param obj - Parsed JSON object with type 'neuron-topic'
 * @returns Validated NeuronTopicService
 * @throws TopicError NEURON-TOPIC-001 if required fields are missing or invalid
 */
export function parseNeuronTopicService(obj: Record<string, unknown>): NeuronTopicService {
  if (obj['type'] !== 'neuron-topic') {
    throw invalidTopicRef(
      `Expected service type "neuron-topic", got "${String(obj['type'])}"`,
    );
  }

  const name = requireString(obj, 'name', 'NeuronTopicService');
  const version = requireString(obj, 'version', 'NeuronTopicService');
  const channel = requireString(obj, 'channel', 'NeuronTopicService');
  const transport = requireString(obj, 'transport', 'NeuronTopicService');
  const anchor = requireString(obj, 'anchor', 'NeuronTopicService');

  if (!isValidChannelRole(channel)) {
    throw invalidTopicRef(
      `Invalid channel role in NeuronTopicService: "${channel}"`,
    );
  }

  if (!isValidBackendKind(transport)) {
    throw invalidTopicRef(
      `Invalid transport kind in NeuronTopicService: "${transport}"`,
    );
  }

  const config = obj['config'];
  if (config === undefined || config === null || typeof config !== 'object') {
    throw invalidTopicRef('NeuronTopicService requires a "config" object');
  }

  const endpoint = typeof obj['endpoint'] === 'string' ? obj['endpoint'] : undefined;

  return {
    type: 'neuron-topic',
    name,
    endpoint,
    version,
    channel,
    transport,
    anchor,
    config: config as TransportConfig,
  };
}

/**
 * Parse a NeuronP2PExchangeService from a JSON object.
 *
 * FR-T17: Validates all required fields.
 *
 * @param obj - Parsed JSON object with type 'neuron-p2p-exchange'
 * @returns Validated NeuronP2PExchangeService
 * @throws TopicError NEURON-TOPIC-001 if required fields are missing or invalid
 */
export function parseNeuronP2PExchangeService(
  obj: Record<string, unknown>,
): NeuronP2PExchangeService {
  if (obj['type'] !== 'neuron-p2p-exchange') {
    throw invalidTopicRef(
      `Expected service type "neuron-p2p-exchange", got "${String(obj['type'])}"`,
    );
  }

  const name = requireString(obj, 'name', 'NeuronP2PExchangeService');
  const version = requireString(obj, 'version', 'NeuronP2PExchangeService');
  const peerID = requireString(obj, 'peerID', 'NeuronP2PExchangeService');
  const protocol = requireString(obj, 'protocol', 'NeuronP2PExchangeService');
  const topicRef = requireString(obj, 'topicRef', 'NeuronP2PExchangeService');

  if (!protocol.startsWith('/')) {
    throw invalidTopicRef(
      `NeuronP2PExchangeService protocol must start with "/", got "${protocol}"`,
    );
  }

  if (!isValidChannelRole(topicRef)) {
    throw invalidTopicRef(
      `Invalid topicRef in NeuronP2PExchangeService: "${topicRef}". Must be a valid channel role`,
    );
  }

  return {
    type: 'neuron-p2p-exchange',
    name,
    version,
    peerID,
    protocol,
    topicRef,
  };
}

/**
 * Parse all Neuron services from an agentURI JSON document.
 *
 * FR-T09: Extracts neuron-topic and neuron-p2p-exchange services from the
 * services array. Other service types are ignored (forward-compatible).
 *
 * @param json - Raw JSON string of the agentURI document
 * @returns Tuple of [topic services, p2p exchange services]
 * @throws TopicError NEURON-TOPIC-001 if parsing fails
 */
export function parseAgentURIServices(
  json: string,
): [NeuronTopicService[], NeuronP2PExchangeService[]] {
  let doc: unknown;
  try {
    doc = JSON.parse(json) as unknown;
  } catch (e) {
    const cause = e instanceof Error ? e : new Error(String(e));
    throw invalidTopicRef('Failed to parse agentURI JSON', cause);
  }

  if (typeof doc !== 'object' || doc === null) {
    throw invalidTopicRef('agentURI document must be a JSON object');
  }

  const docObj = doc as Record<string, unknown>;
  const services = docObj['services'];

  if (!Array.isArray(services)) {
    throw invalidTopicRef('agentURI document must have a "services" array');
  }

  const topics: NeuronTopicService[] = [];
  const p2pExchanges: NeuronP2PExchangeService[] = [];

  for (const svc of services) {
    if (typeof svc !== 'object' || svc === null) {
      continue;
    }

    const svcObj = svc as Record<string, unknown>;
    const svcType = svcObj['type'];

    if (svcType === 'neuron-topic') {
      topics.push(parseNeuronTopicService(svcObj));
    } else if (svcType === 'neuron-p2p-exchange') {
      p2pExchanges.push(parseNeuronP2PExchangeService(svcObj));
    }
    // Other types are ignored (forward-compatible)
  }

  return [topics, p2pExchanges];
}

/**
 * Validate cross-references between P2P exchange services and topic services.
 *
 * FR-T18, SC-T10: Every topicRef in a NeuronP2PExchangeService MUST resolve
 * to an existing NeuronTopicService name in the same agentURI document.
 *
 * @param topics - Parsed topic services from the agentURI
 * @param p2p - Parsed P2P exchange services
 * @throws TopicError NEURON-TOPIC-001 if any cross-reference is broken
 */
export function validateCrossReferences(
  topics: readonly NeuronTopicService[],
  p2p: readonly NeuronP2PExchangeService[],
): void {
  const topicNames = new Set(topics.map(t => t.name));

  for (const svc of p2p) {
    if (!topicNames.has(svc.topicRef)) {
      throw invalidTopicRef(
        `Broken topicRef in NeuronP2PExchangeService "${svc.name}": ` +
        `topicRef "${svc.topicRef}" does not match any neuron-topic service name. ` +
        `Available names: [${Array.from(topicNames).join(', ')}]`,
      );
    }
  }
}

/**
 * Extract a TopicRef from a NeuronTopicService.
 *
 * FR-T09: The locator is extracted from the config based on transport:
 * - hcs: config.topicId
 * - erc-log: config.chainId:config.contractAddress
 * - kafka: config.topicName
 * - custom: adapter-defined extraction (falls back to endpoint)
 *
 * @param svc - A parsed neuron-topic service
 * @returns TopicRef instance
 * @throws TopicError NEURON-TOPIC-001 if extraction fails
 */
export function extractTopicRef(svc: NeuronTopicService): TopicRef {
  const config = svc.config as Record<string, unknown>;

  switch (svc.transport) {
    case 'hcs': {
      const topicId = config['topicId'];
      if (typeof topicId !== 'string' || topicId.length === 0) {
        throw invalidTopicRef('HCS config missing required field "topicId"');
      }
      return TopicRef.create('hcs', topicId);
    }
    case 'erc-log': {
      const chainId = config['chainId'];
      const contractAddress = config['contractAddress'];
      if (typeof chainId !== 'number' || typeof contractAddress !== 'string') {
        throw invalidTopicRef(
          'ERC-log config missing required fields "chainId" and/or "contractAddress"',
        );
      }
      return TopicRef.create('erc-log', `${chainId.toString()}:${contractAddress}`);
    }
    case 'kafka': {
      const topicName = config['topicName'];
      if (typeof topicName !== 'string' || topicName.length === 0) {
        throw invalidTopicRef('Kafka config missing required field "topicName"');
      }
      return TopicRef.create('kafka', topicName);
    }
    default: {
      // Custom transport: try config-based or fall back to endpoint
      if (svc.endpoint !== undefined && svc.endpoint.length > 0) {
        return TopicRef.create(svc.transport as BackendKind, svc.endpoint);
      }
      throw invalidTopicRef(
        `Cannot extract TopicRef from custom transport "${svc.transport}" without endpoint`,
      );
    }
  }
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/**
 * Require a string field from a parsed JSON object.
 *
 * @param obj - The object to read from
 * @param field - The field name
 * @param context - Error context (e.g., 'NeuronTopicService')
 * @returns The string value
 * @throws TopicError NEURON-TOPIC-001 if field is missing or not a string
 */
function requireString(
  obj: Record<string, unknown>,
  field: string,
  context: string,
): string {
  const value = obj[field];
  if (typeof value !== 'string' || value.length === 0) {
    throw invalidTopicRef(
      `${context} requires non-empty string field "${field}"`,
    );
  }
  return value;
}
